src/paperoni/discovery/miniconf.py

```diff
@@ -122,6 +125,9 @@ def expand_base(uri):
             links.add(Link(type="pdf", link=url))
         if url := expand_base(data.get("virtualsite_url")):
             links.add(Link(type="abstract", link=url))
+        links = sorted(
+            links, key=lambda x: {"pdf": 0, "abstract": 1}.get(x.type, math.inf)
+        )
```

Comment:
sorted() keeps original ordering in case of equal keys, but if the set is not deterministically ordered, the result won't be deterministic either. Better just use (type, link) as a key.

Reply:
I say we leave the links unsorted in the db, it's easy enough to sort them at the final step.